File: source/window/GLContext.java
package window;

import math.Float4;
import math.Int2;
import org.lwjgl.opengl.GL;

import static org.lwjgl.opengl.GL33.*;

public final class GLContext {
    public GLContext() {
        GL.createCapabilities();
    }

    public void destroy() {
        GL.setCapabilities(null);
    }

    public void setViewport(Int2 size) {
        setViewport(new Int2(), size);
    }

    public void setViewport(Int2 pos, Int2 size) {
        glViewport(pos.x, pos.y, size.x, size.y);
    }

    public void setClearColor(Float4 color) {
        glClearColor(color.x, color.y, color.z, color.w);
    }

    public void clear(boolean depth) {
        glClear(GL_COLOR_BUFFER_BIT | (depth ? GL_DEPTH_BUFFER_BIT : 0));
    }

    public void setDepthTest(boolean enabled) {
        if (enabled) {
            glEnable(GL_DEPTH_TEST);
        }
        else {
            glDisable(GL_DEPTH_TEST);
        }
    }

    public void setWireframe(boolean wireframe) {
        glPolygonMode(GL_FRONT_AND_BACK, wireframe ? GL_LINE : GL_FILL);
    }

    public void setCull(boolean enabled, boolean cullBack) {
        if (enabled) {
            glEnable(GL_CULL_FACE);
            glCullFace(cullBack ? GL_BACK : GL_FRONT);
        }
        else {
            glDisable(GL_CULL_FACE);
        }
    }
}

File: source/window/GLContext.kt
package window

import math.Float4
import math.Int2
import org.lwjgl.opengl.GL
import org.lwjgl.opengl.GL11

class GLContext {
    init {
        GL.createCapabilities()
    }

    fun destroy() {
        GL.setCapabilities(null)
    }

    fun setViewport(size: Int2?) {
        setViewport(Int2(), size)
    }

    fun setViewport(pos: Int2, size: Int2?) {
        GL11.glViewport(pos.x, pos.y, size!!.x, size.y)
    }

    fun setClearColor(color: Float4) {
        GL11.glClearColor(color.x, color.y, color.z, color.w)
    }

    fun clear(depth: Boolean) {
        GL11.glClear(GL11.GL_COLOR_BUFFER_BIT or if (depth) GL11.GL_DEPTH_BUFFER_BIT else 0)
    }

    fun setDepthTest(enabled: Boolean) {
        if (enabled) {
            GL11.glEnable(GL11.GL_DEPTH_TEST)
        } else {
            GL11.glDisable(GL11.GL_DEPTH_TEST)
        }
    }

    fun setWireframe(wireframe: Boolean) {
        GL11.glPolygonMode(GL11.GL_FRONT_AND_BACK, if (wireframe) GL11.GL_LINE else GL11.GL_FILL)
    }

    fun setCull(enabled: Boolean, cullBack: Boolean) {
        if (enabled) {
            GL11.glEnable(GL11.GL_CULL_FACE)
            GL11.glCullFace(if (cullBack) GL11.GL_BACK else GL11.GL_FRONT)
        } else {
            GL11.glDisable(GL11.GL_CULL_FACE)
        }
    }
}